bson/primitive/objectid.go

```diff
@@ -36,15 +36,26 @@ var processUnique = processUniqueBytes()
 
 // NewObjectID generates a new ObjectID.
 func NewObjectID() ObjectID {
+	return NewObjectIDFromTime(time.Now())
+}
+
+// NewObjectIDFromTime generates a new ObjectID based on the given time.
+func NewObjectIDFromTime(generationTime time.Time) ObjectID {
 	var b [12]byte
 
-	binary.BigEndian.PutUint32(b[0:4], uint32(time.Now().Unix()))
+	binary.BigEndian.PutUint32(b[0:4], uint32(generationTime.Unix()))
 	copy(b[4:9], processUnique[:])
 	putUint24(b[9:12], atomic.AddUint32(&objectIDCounter, 1))
 
 	return b
 }
 
+// GenerationTime extracts the time part of the ObjectId.
+func (id ObjectID) GenerationTime() time.Time {
```

Comment:
This method should be called `Timestamp` since that's what the value is.

Reply:
Renamed